### ISSUE TITLE
Ubuntu non-LTS non-support

### ISSUE THREAD

**Issue by oerdnj** [oerdnj/deb.sury.org#1662]
After some thought, these repositories will not support the non-LTS Ubuntu versions. It requires an extra effort to add all the packages to the new Ubuntu release, and it does consume computing cycles every time there's a new PHP (or any other package updates).

The release cycle on non-LTS Ubuntu release is just 9 months this feels like a waste of just everything. There also doesn't seem to be a very high demand for these packages - only a handful of people wrote here, to mail or on Twitter asking for Ubuntu Impish support.

Ubuntu 21.10 Impish Indri will be the last non-LTS release supported.

**Comment by baldhead** [comment]
Make sense, though I guess a lot of people will be disappointed - including me :) Always appreciated your awesome work. Will you update this topic if you will do this version - if not, I'll revert my OS to 21.04 (which also has a 9 months EOL cycle) Thanks and either way ... looking forward to the 22.04 version then!

**Comment by oerdnj** [comment]
Here's a question to start a discussion.

What is the particular use case that makes you run the latest Ubuntu release as a web server.

I can understand the desktop as development machine where you want to run the latest desktop environment and all the PHP versions.

**Comment by jacobseated** [comment]
Possibly because I have a Ryzen laptop with poor hardware support, and I am desperately waiting for better support for AMD graphics in VirtualBox.

I can not run hardware accelerated software in my Windows VM currently, and I suspect the issue may be Kernel related. I have also had different issues in the past that was resolved by updating the Kernel.

And, it seems we now need signed Ubuntu Kernels because of modern bios configurations. Possibly this is related to secure boot or something — and Kernels are only signed on official releases afaik

But, other then that, I actually prefer to run the latest software on my main system, and then Snaps or Appimages when available. Interestingly, FFMpeg also broke, so I had to compile a standalone to get video conversion working, but that was in the beta, and I have not checked if the dependencies has been fixed in the final 21.10 release. 

I wanted to try and move my servers to a VM for some time, and now I finally got shared folders working; I figure, why not give it a try for now, because it seem to enable me to test the latest software more freely :-)

**Comment by oerdnj** [comment]
For apache2, there's ppa:ondrej/apache2 and for nginx, there are ppa:ondrej/nginx and ppa:ondrej/nginx-mainline.

**Comment by stadja** [comment]
Well in my case it is kind of a silly reason and it's not for a production server, but for my dev machine.

I am a freelance. And I have many different clients with many different server specs: i can't have a différent VM for all my clients.
One of my client is stuck with PHP7.4 right now and can't upgrade to PHP8.*

On my dev machine, a long time ago, I really don't know why, I switched my do-release-upgrade from "lts" to "normal" (i know, it's bad.)
And this weekend, I was not thinking about it and made the upgrade from 21.04 to 21.10... and know, I can't dev for my client because I don't have the right PHP)

I made plenty of mistake (lts to normal and upgrade without thinking), that is why I am looking for a way to be able to keep working for this client.... the other way will be te reinstall a lts ubuntu on my machine. But I kind of fear this prospect (will I lose a lot of time ? will I lose some strange configurations I did 5 years ago ? that kind of things)

**Comment by oerdnj** [comment]
FTR for my $dayjob I have a beefier machine in the "basement" with Proxmox installed and have multiple operating systems and versions in separate VMs for testing compatibility when there's a problem in the CI.

I found out that long-term it's better to separate the development environment from the "desktop" machine, because they usually have different lifecycles.

**Comment by stadja** [comment]
So, as right now I am stuck with Impish and I really needed a php 7.4 working, this is what I did:

I added the repo
```
sudo add-apt-repository ppa:ondrej/php
```
edited the sources.list.d
```
sudo nano /etc/apt/sources.list.d/ondrej-ubuntu-php-impish.list
```

I changed the `impish main` to `hirsute main` and now it's okay, I can install and use my php7.4
```
sudo apt-get install php7.4 php7.4-cli php7.4-common php7.4-json php7.4-opcache php7.4-mysql php7.4-mbstring  php7.4-zip php7.4-fpm php7.4-intl php7.4-simplexml
sudo a2dismod php8.0
sudo a2enmod php7.4
sudo update-alternatives --set php /usr/bin/php7.4
sudo update-alternatives --set phar /usr/bin/phar7.4
sudo update-alternatives --set phar.phar /usr/bin/phar.phar7.4
sudo update-alternatives --set phpize /usr/bin/phpize7.4
sudo update-alternatives --set php-config /usr/bin/php-config7.4
```

I hope it helps someone somewhere :)

(next time I won't upgrade that thoughtlessly... you can think that 15 years using linux teaches you that ... but nop)

**Comment by jacobseated** [comment]
> Then simply use some Docker based stuff like https://github.com/drud/ddev that helps you to use different PHP versions per project in a very simple manner

Exactly. I am not too familiar with docker, so I simply installed the LTS in a VM and then installed multiple PHP versions via ppa:ondrej/php — we just got to link them in the Apache configuration and enable the individual configuration files. I just love the convenience of Apache :-P 

Typically I use the latest PHP-FPM with fcgi as a global configuration, and then I override the default on a per VHOST basis as needed. E.g. Inside the VirtualHost block, and outside of directory blocks:
`
include /etc/apache2/conf-available/php7.4-fpm.conf`

Of course, php7.4-fpm should **not** be enabled via "a2enconf", but rather by manually loading it for the VHOST that needs it. Hope that makes sense to some. Have no idea about nginx unfortunately, because I have not used that for a long time.

The cool thing I found out is that I can just share my hosts **/var/www/** folder with the VirtualBox, and than add Apache (the www-data user) to the vboxsf group, which then allows me to run headless and still edit files from my host.

**Comment by loxK** [comment]
@oerdnj I'm interested as well. For some PHP extensions and for switching between 5.6, 7.4 and 8.0 on my development workstation that I like keeping on the edge regarding Ubuntu versions. I am worried that it is such a hard work for you ... 

The most annoying is bringing that issue in between two rolling releases, but if you can keep the hard work until next LTS and then decide to only support LTS release, I must admit that I personally would be very thankful ...

**Comment by oerdnj** [comment]
FTR I will do the Ubuntu 21.10 as the last non-LTS release and after Ubuntu 22.04, only LTS releases will be supported by this repository.

**Comment by Ayesh** [comment]
I upgraded to 21.10, and will be staying with 22.04 LTS. Thank you @oerdnj for providing a smooth path to 22.04.

**Comment by enderandrew** [comment]
I don't know if this comment is best served in #1439 or here.

I've got a VM where I'm putting together a few websites for some local Masonic fraternity websites. I wanted to test on the bleeding edge for security and performance reasons. All too-often there are zero-day exploits and I wanted to see if I can get everything working on the bleeding edge if possible. I was going to test with PHP 8 but seeing the RC out and that you have  8.1 RC4 in your repository, I decided to add it.

However, I've started with Debian Bookworm, and your PHP repository doesn't have a Bookworm version. Perhaps foolishly I added it none the less and have the 8.1 Bullseye packages installed on top of Bookworm. Is there anything I can do to fix this?

dpkg: error processing archive /var/cache/apt/archives/libpcre2-posix2_10.39-1+0~20211102.13+debian11~1.gbp2315ad_amd64.deb (--unpack):
 trying to overwrite '/usr/lib/x86_64-linux-gnu/libpcre2-posix.so.3.0.1', which is also in package libpcre2-posix3:amd64 10.39-2
Errors were encountered while processing:
 /var/cache/apt/archives/libpcre2-posix2_
 
You asked:

"What is the particular use case that makes you run the latest Ubuntu release as a web server."

For a production webserver at work I'd never run bleeding edge. But I'm an amateur web dev and all too often I see smaller sites easily compromised. There is some security benefit to living on the bleeding edge to avoid newer exploits.

Given that reason, would you consider adding Bookwork packages?

**Comment by Muffinman** [comment]
There's not really a valid argument here (in my opinion). Bleeding edge is _more_ likely to have unforeseen bugs/exploits, and any security issues at the distro level would be fixed in the LTS at the same time.

In the 14 years I've been doing commercial PHP development, I've never seen a site exploited due to an exploit in PHP itself, it's always been implementation flaws.

I'm not saying it won't happen, but it's certainly not the main attack vector you should be concerned about.

As for the error, I'm not sure I'm afraid.

**Comment by oerdnj** [comment]
> Given that reason, would you consider adding Bookwork packages?

No.

Also this issue is about **Ubuntu** not Debian.

**Comment by oerdnj** [comment]
> Bleeding edge is more likely to have unforeseen bugs/exploits, and any security issues at the distro level would be fixed in the LTS at the same time.

It's the other way around. Release in development has no ETA when fix for security issue will hit, it might be blocked by transition or other bug.  You should not run development version unless you really know what you are doing and why are you doing that.

**Comment by oerdnj** [comment]
That said I am now going to delete all this as it doesn't belong to this issue.

**Comment by domis86** [comment]
@loxK @gjm @jacobseated @stadja Please dont torture yourself and just use docker to manage multiple projects on same machine - then each such project can have separate OS version, php version, apache version etc :) 

Alternatively, as was also mentioned: VM. Like Vagrant+Virtualbox etc.

Additional bonus which you get from above approaches is that you can always simply "delete" your docker container (or VM) and rebuild it from scratch - without affecting your computer. Is a lot less pain when for example you need to upgrade something (Ubuntu version etc) in one of your projects.

**Comment by hswong3i** [comment]
@oerdnj Thank you very much for maintenance this very useful PHP repo for many years, and I had been using it for my Ansible Role and Docker image for few years, too:
* https://github.com/alvistack/ansible-role-php
* https://github.com/alvistack/docker-php
  * https://hub.docker.com/r/alvistack/php-8.0/tags
  * https://hub.docker.com/r/alvistack/php-7.4/tags
  * https://hub.docker.com/r/alvistack/php-7.3/tags

Also sharing some though when maintenance my repo (https://build.opensuse.org/project/show/home:alvistack) for latest Ansible / Kubernetes / CRI-O / Podman / etc:
* For LTS release I keep a `N / N -1` policy
  * e.g. Ubuntu 20.04 / 18.04, CentOS 8 / 7, openSUSE Leap 15.3 / 15.2
* For point release I keep sync with its `EOL`
  * e.g. Only support Ubuntu 21.10 / 21.04 (to be remove on 2022-01)
  * Remove Ubuntu 20.10 support (EOL since 2021-07)
* For rolling release I simply keep it as technical preview (i.e. allow failure)
  * e.g. Debian Testing, Fedora Rawhide, openSUSE Tumbleweed
  * Keeping it active so whenever new point / LTS comes upgrade could be much simpler
* For each package I keep multiple version with policy `MAJOR x1 / MINOR x2 / PATCH x3`
  * e.g. For podman I remove all v2.y.z release, just keep recent v3.3.z and v3.4.z releases

P.S. I saw that you need to manually maintenance https://packages.sury.org/ for Debian based packages which could be very time consuming; working with https://build.opensuse.org/ for all supported OS seems much simple then using Ubuntu PPA (for Ubuntu only):
* It could support both Debian / Ubuntu within single repo
* Support automatic rebuild triggered by corresponding dependencies got upgraded
* I simply do a quick local build with my Ubuntu 21.10, once it looks good and pushed to OBS, OBS will handle all OS build for me at once
* If I need to debug the packaging, I could also have an isolated local build environment for Debian 10 / 11 / Ubuntu 18.04 / 20.04 / 21.04 / etc, on top of my daily used Ubuntu 21.10

Again, thank you very much for maintenance this repo ;-)

**Comment by Jimbolino** [comment]
For me the pressure to use these packages increases when the following conditions are met:
- The latest ubuntu LTS is getting too old to daily drive
- The new ubuntu LTS is shipping soon with a new version of PHP

Since PHP releases +- once a year, and ubuntu LTS once every 2 years. I predict that in 2 years, when ubuntu 23.10 is released the same discussing will arise :)

**Comment by jadeops** [comment]
Any chance of supporting the Ubuntu 22.10 release (kinetic) ? beta is available now. How about supporting the versions only supported by PHP upstream? - https://www.php.net/supported-versions.php

I use Debian on all servers but use Xubuntu for desktop (dev machine). Debian is too stale for desktop use.

I am looking for
- php8.1
- php8.2(upcoming) to migrate code

If not, is there any cheatsheet/scripts to build this ourselves?

**Comment by Etukakpan** [comment]
Hello oerdnj please i keep having this issue when i try to install the  Install PHP 7.4 on my ubuntu 20.0 please help me out i have a deadline and it all on me installing this on my system
![Screen Shot 2022-10-18 at 3 58 47 AM](https://user-images.githubusercontent.com/84282974/197574011-ce3923cc-0dd1-4a5b-a9fc-7fc95f01910b.png)
4

**Comment by oerdnj** [comment]
> Hello oerdnj please i keep having this issue when i try to install the Install PHP 7.4 on my ubuntu 20.0 please help me out i have a deadline and it all on me installing this on my system
> ![Screen Shot 2022-10-18 at 3 58 47 AM](https://user-images.githubusercontent.com/84282974/197574011-ce3923cc-0dd1-4a5b-a9fc-7fc95f01910b.png)
> 4

Don’t pollute random issues with request that are off-topic. I am going to delete your comment (and mine).

**Comment by oerdnj** [comment]
> Any chance of supporting the Ubuntu 22.10 release (kinetic) ? beta is available now. How about supporting the versions only supported by PHP upstream? - https://www.php.net/supported-versions.php
> 
> I use Debian on all servers but use Xubuntu for desktop (dev machine). Debian is too stale for desktop use.
> 
> I am looking for
> 
> * php8.1
> * php8.2(upcoming) to migrate code
> 
> If not, is there any cheatsheet/scripts to build this ourselves?

Checkout the git repository, switch to debian/main/8.2 branch and run git-buildpackage - it will guide you through missing build-deps, etc…

**Comment by jefhar** [comment]
> these repositories will not support the non-LTS Ubuntu versions.

Would it be asking too much for you add that to the `README.md` and the https://deb.sury.org/ page so that they are visible in more places that just this issue?

**Comment by str** [comment]
What about a package management that does not depend on Ubuntu's version, like snaps or flatpack?

I do not like snaps, and have never tried flatpack, but testing the same project with multiple PHP versions with just changing a line in an .htaccess is soooo convenient, I'm about to downgrade from 23.04 to previous LTS. I'm THAT desperate.

I was thinking on finally jump to Debian, but I see the the Debian repo is not available for testing/rolling, right?

So, again...  have you ever think of a package management that does not depend on Ubuntu's version, like snaps or flatpack?

**Comment by jacobseated** [comment]
Just use Docker, devenv or something similar, it is way better than running
stuff directly on your main os for numerous of reasons.

**Comment by davidbuzz** [comment]
just throwing a note out that I've upgraded my multi-php dev box jammy to lunar[23.04] and find that continuing to use the jammy ppa's "works for me" with php7 and php8 on apache:

$ grep ondrej /etc/apt/sources.list.d/*.list
/etc/apt/sources.list.d/ondrej-ubuntu-apache2-kinetic.list:deb https://ppa.launchpadcontent.net/ondrej/apache2/ubuntu/ jammy main # disabled on upgrade to lunar
/etc/apt/sources.list.d/ondrej-ubuntu-php-kinetic.list:deb https://ppa.launchpadcontent.net/ondrej/php/ubuntu/ jammy main # disabled on upgrade to lunar
$ uname -a
Linux hostname 6.2.0-20-generic #20-Ubuntu SMP PREEMPT_DYNAMIC Thu Apr  6 07:48:48 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=23.04
DISTRIB_CODENAME=lunar
DISTRIB_DESCRIPTION="Ubuntu 23.04"
$ php7.4 --version
PHP 7.4.33 (cli) (built: Feb 14 2023 18:31:54) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Xdebug v3.1.6, Copyright (c) 2002-2022, by Derick Rethans
$ php8.0 --version
PHP 8.0.28 (cli) (built: Feb 14 2023 18:33:29) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.28, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.28, Copyright (c), by Zend Technologies
    with Xdebug v3.2.0, Copyright (c) 2002-2022, by Derick Rethans



not my first rodeo. :-)

**Comment by Firhan384** [comment]
i got error after install add-apt-repository ppa:ondrej/php 
error :
The repository 'https://ppa.launchpadcontent.net/ondrej/php/ubuntu lunar Release' does not have a Release file.

uname -a
Linux firhan 6.2.0-20-generic #20-Ubuntu SMP PREEMPT_DYNAMIC Thu Apr  6 07:48:48 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=23.04
DISTRIB_CODENAME=lunar
DISTRIB_DESCRIPTION="Ubuntu 23.04"

**Comment by loxK** [comment]
@Firhan384 read the first sentence of this issue:

> After some thought, these repositories will not support the non-LTS Ubuntu versions.

**Comment by str** [comment]
> I'm about to downgrade from 23.04 to previous LTS. I'm THAT desperate.

And I ended downgrading to LTS, sadly 😭

**Comment by TomasVotruba** [comment]
This is not recommended! But I need legacy PHP versions locally, while keeping my stack up-to-date as possible. I do upgrade legacy projects for a living :)


Running Ubuntu 25.04 since yesterday here. Only available PHP version on this version is PHP 8.4 (even I don't run it on any project yet).


This is a hack to get old PHP versions on Ubuntu 25.04:

```bash
# edit sources file
nano /etc/apt/sources.list.d/ondrej-ubuntu-php-plucky.sources

# change "suites: x" to "suites: noble"

sudo apt-get update
```

The install as usuall.
